pallets/pallet-did-lookup/src/migrations.rs

```diff
@@ -16,28 +16,124 @@
 	Blake2_128Concat,
 };
-use sp_std::{marker::PhantomData, vec};
-
-#[cfg(feature = "try-runtime")]
-use frame_support::traits::OnRuntimeUpgradeHelpersExt;
-#[cfg(feature = "try-runtime")]
-use sp_runtime::SaturatedConversion;
+use sp_std::{marker::PhantomData, vec::Vec};
 
+/// Keytype changed from `AccountId` to `LinkableAccountId` changed in V3
 #[storage_alias]
 type ConnectedDids<T: Config> = StorageMap<Pallet<T>, Blake2_128Concat, AccountIdOf<T>, ConnectionRecordOf<T>>;
+/// Second keytype changed from `AccountId` to `LinkableAccountId` changed in V3
 #[storage_alias]
 type ConnectedAccounts<T: Config> =
 	StorageDoubleMap<Pallet<T>, Blake2_128Concat, DidIdentifierOf<T>, Blake2_128Concat, AccountIdOf<T>, ()>;
 
+/// Migrate the `ConnectedDids` and `ConnectedAccounts` key types for a given
+/// `AccountId`.
+pub(crate) fn do_migrate_account_id<T: Config>(
+	account_id: AccountIdOf<T>,
+	linkable_account: LinkableAccountId,
+) -> Option<DidIdentifierOf<T>> {
+	ConnectedDids::<T>::take(&account_id).map(|did_record| {
+		ConnectedDidsV2::<T>::insert(&linkable_account, did_record.clone());
+		if let Some(v) = ConnectedAccounts::<T>::take(&did_record.did, &account_id) {
+			ConnectedAccountsV2::<T>::insert(&did_record.did, linkable_account, v);
+		}
+		did_record.did
+	})
+}
+
+/// Sanity check that both new typed storage maps `ConnectedDids`,
+/// `ConnectedAccounts` have as many keys as their old counter parts with
+/// different key types.
+pub(crate) fn check_storage_size<T: Config>() -> bool {
+	ConnectedAccounts::<T>::iter_keys().count() == ConnectedAccountsV2::<T>::iter_keys().count()
+		&& ConnectedDids::<T>::iter_keys().count() == ConnectedDidsV2::<T>::iter_keys().count()
+}
+
+/// Iterates over both old typed storage maps `ConnectedDids`,
+/// `ConnectedAccounts` and checks whether any raw storage key still exists in
+/// the low level storage.
+pub(crate) fn do_verify_migration<T: Config>() -> bool {
+	check_did_migration::<T>(None).is_empty() && check_account_migration::<T>(None).is_empty()
+}
+
+/// Iterates over old connected did storage map and checks whether any raw key
+/// still exists in the low level storage.
+///
+/// Since the new `ConnectedDidsV2` and old `ConnectedDids` typed storage maps
+/// have the same pallet and storage prefixes, both result in the same final
+/// storage map key. For some reason, keys in the new map can still be iterated
+/// over in the old one. E.g., the new keytype `LinkableAccountId` can be
+/// decoded into the old one `AccountId32` such that both maps have the same
+/// number of keys despite killing every old key during the migration.
+///
+/// However, we can check the old raw keys which should not exist in storage
+/// after migrating, e.g. `unhashed::exists(old_raw_key)` is expected to be
+/// false.
+pub(crate) fn check_did_migration<T: Config>(maybe_last_key: Option<Vec<u8>>) -> Vec<(AccountIdOf<T>, Vec<u8>)> {
+	if let Some(last_key) = maybe_last_key {
```

Comment:
Because we are reading from and writing to the same map (this is still my assumption, please correct me if I am wrong), is this safe to do?
Wouldn't it be enough to always start from the beginning and try to decode each key returned by the iterator as a `LinkableAccountId`? If any decoding fails, it means that there are still `AccountId` among the keys.

Reply:
I am not sure how the iterator works, wheather it keeps some order for the items. In any case, the `verify_migration` does not take any parameter (yet?), so that means that we always start from the beginning anyway, right? Why having this caching logic?

Reply:
Please ignore the part about the same map, but still the `maybe_last_key` is never passed from the extrinsic, so where can it actually be used besides in tests? If it can't be used, why do we test it?

Reply:
My idea was to expose a runtime API call for these both functions. See above in the PR description

> I envision to use a runtime API call which returns a `Vec<AccountId>` of unmigrated account ids between step 1. and 2. This call could use `pallet_did_lookup::migrations::check_did_migration` and `pallet_did_lookup::migrations::check_account_migration`, so everything's already there. However, we could also simply query and temporarily store the account ids shortly before the runtime upgrade via `api.query.connectedDids.entries()` and `api.query.connectedAccounts.entries()`.